libs/ml/tests/ml/ops/squeeze.cpp

```diff
@@ -0,0 +1,182 @@
+
+  TensorType data = TensorType::FromString("1, 2, 4, 8, 100, 1000");
+  data.Reshape({6, 1});
+
+  fetch::ml::ops::Squeeze<TypeParam> op;
+
+  TypeParam prediction(op.ComputeOutputShape({std::make_shared<const TensorType>(data)}));
+  op.Forward({std::make_shared<const TensorType>(data)}, prediction);
+
+  ASSERT_EQ(prediction.shape().size(), 1);
+  ASSERT_EQ(prediction.shape().at(0), 6);
+
+  ASSERT_TRUE(prediction.AllClose(data, fetch::math::function_tolerance<DataType>(),
+                                  fetch::math::function_tolerance<DataType>()));
+}
+
+TYPED_TEST(SqueezeTest, backward_test)
+{
+  using TensorType = TypeParam;
+  using DataType   = typename TypeParam::Type;
+
+  TensorType data = TensorType::FromString("1, -2, 4, -10, 100");
+  data.Reshape({1, 5});
+  TensorType error = TensorType::FromString("1, 1, 1, 2, 0");
+  error.Reshape({5});
+
+  auto shape = error.shape();
+
+  fetch::ml::ops::Squeeze<TypeParam> op;
+
+  std::vector<TensorType> prediction =
+      op.Backward({std::make_shared<const TensorType>(data)}, error);
+
+  ASSERT_EQ(prediction.at(0).shape().size(), 2);
+  ASSERT_EQ(prediction.at(0).shape().at(0), 1);
+  ASSERT_EQ(prediction.at(0).shape().at(1), 5);
+
+  ASSERT_TRUE(prediction.at(0).AllClose(error, fetch::math::function_tolerance<DataType>(),
+                                        fetch::math::function_tolerance<DataType>()));
+  fetch::math::state_clear<DataType>();
+}
+
+TYPED_TEST(SqueezeTest, saveparams_test)
+{
+  using TensorType    = TypeParam;
+  using DataType      = typename TypeParam::Type;
+  using VecTensorType = typename fetch::ml::ops::Ops<TensorType>::VecTensorType;
+  using SPType        = typename fetch::ml::ops::Squeeze<TensorType>::SPType;
+  using OpType        = fetch::ml::ops::Squeeze<TensorType>;
+
+  TensorType data = TensorType::FromString("1, 2, 4, 8, 100, 1000");
+  data.Reshape({6, 1});
+
+  OpType op;
+
+  TensorType    prediction(op.ComputeOutputShape({std::make_shared<const TensorType>(data)}));
+  VecTensorType vec_data({std::make_shared<const TensorType>(data)});
+
+  op.Forward(vec_data, prediction);
+
+  // extract saveparams
+  std::shared_ptr<fetch::ml::OpsSaveableParams> sp = op.GetOpSaveableParams();
+
+  // downcast to correct type
+  auto dsp = std::static_pointer_cast<SPType>(sp);
+
+  // serialize
+  fetch::serializers::MsgPackSerializer b;
+  b << *dsp;
+
+  // deserialize
+  b.seek(0);
+  auto dsp2 = std::make_shared<SPType>();
+  b >> *dsp2;
+
+  // rebuild node
+  OpType new_op(*dsp2);
+
+  // check that new predictions match the old
+  TensorType new_prediction(op.ComputeOutputShape({std::make_shared<const TensorType>(data)}));
+  new_op.Forward(vec_data, new_prediction);
+
+  // test correct values
+  EXPECT_TRUE(
+      new_prediction.AllClose(prediction, static_cast<DataType>(0), static_cast<DataType>(0)));
+  fetch::math::state_clear<DataType>();
+}
+
+TYPED_TEST(SqueezeTest, saveparams_backward_test)
```

Comment:
please add a test that also adds the op to a graph and then serialises and deserialises the graph.

I suspect it will fail because you will need to add a little bit to the graph_builder code

Reply:
You were right, it's fixed now.